### ISSUE TITLE
Investigate allowing mathematic notation in SiteNow

### ISSUE THREAD

**Issue by briand44** [uiowa/uiowa801#9753]
# Motivation

We have had multiple requests from CLAS related departments/units to be able to add LaTeX code to SiteNow pages. This is becoming more important as legacy systems are being retired (MyWeb, CLAS personal web page service, etc.).

Latest request (JJ U.):

> I’m deprecating the Apache based personal web page offering I’ve run for CLAS for 20+ years.  Can something like [mathJax] (https://www.mathjax.org/) or equivalent be added to Sitenow?

# Additional Info

@cory-skeers -- related issues:
-  https://github.com/uiowa/uiowa/issues/4152
- https://github.com/uiowa/uiowa/issues/4844

It seems like an investigation to understand our options here would be valuable.

- https://www.drupal.org/project/mathjax
- https://www.drupal.org/node/159031
- https://ckeditor.com/mathtype/, which may be an option. Looking for thoughts from the team. - Alan

**Comment by e-marie-w** [comment]
Want to make visible that there's an upcoming LaTeX accessibility training that mentions using MathJax:  https://accessibility.uiowa.edu/event/36504/0.  While it's focused on ICON and Pressbooks, I was thinking of attending just to learn more about LaTeX/MathJax in general.

**Comment by briand44** [comment]
## Grooming
 
**Quadrant:** 

Complicated

 **Discussion:**
 - Familiarity on campus with MathJax
   - Jonna mentioned Eden J. as someone who could get us in touch with how this was implemented in ICON
   - Eden demoed at a Liasons meeting and that implementation was accessible.
 - ckeditor plugin is interesting as we will want folks to be able to add these mathematical formulas/equations on pages
   - https://ckeditor.com/cke4/addon/mathjax
 - Scope to investigation, with deliverable being summary of what we learned and recommendations for moving forward
 - Should evaluate accessibility and security as well

**Proposed solution or definition of done:**

**Comment by cory-skeers** [comment]
Trying out the drupal MathJax project, and was able to get it up and running within the `filtered_html` text format pretty quickly.

<img width="1026" height="419" alt="Image" src="https://github.com/user-attachments/assets/f3fc1d32-d24c-49ca-b7ba-350800cbfbc8" />

Still exploring, but so far it looks like it is entirely dependent on inserting LaTeX directly into the editor, and filtering it on render. So would work well for those with experience, but it would be brittle and potentially confusing for anyone else editing the page.

Example of the editor from the above screenshot. The default markers for it are `$$` for a separated block and `$` for inline:
<img width="958" height="478" alt="Image" src="https://github.com/user-attachments/assets/4da35de8-f1c2-42ec-91c2-cbec26e08315" />

The only flagged SiteImprove issue I'm getting is an 'S' best practice: https://alfa.siteimprove.com/rules/sia-r86

Because it's loading in the MathJax JS library, there are some built-in options for the end user to interact with it in different ways:

<img width="432" height="229" alt="Image" src="https://github.com/user-attachments/assets/0ac13a6d-ca8d-4e6e-b435-2f63768ba574" />

**Comment by joewhitsitt** [comment]
I think raw in ckeditor is fine as long as there isn't something that is stripped on source toggle or render.

Do we know if the ICON group is displaying that menu of options?

**Comment by cory-skeers** [comment]
> I think raw in ckeditor is fine as long as there isn't something that is stripped on source toggle or render.

So far, I haven't had any issues with it.

<img width="618" height="122" alt="Image" src="https://github.com/user-attachments/assets/8c3b14f2-7ee5-4214-8d68-8bb745eba10e" />
Looks like

```
$$\oint_C \left( \frac{\sum_{k=1}^n \frac{x^k}{k!}}{\sqrt[3]{1 + \int_0^\infty e^{-t^2} dt}} \right) d\mathbf{z} = \begin{pmatrix} \alpha & \beta \\ \gamma & \delta \end{pmatrix}^{-1} \times \prod_{i=1}^m \lim_{n \to \infty} \left( 1 + \frac{1}{n} \right)^n$$
```

in the editor. Lots of slashes and brackets, but ckeditor seems to be leaving it alone alright even with switching back and forth.

The main 'brittleness' then I think is if multiple editors touch a page, it'd be really easy for one not familiar with LaTeX to break something when seeing essentially gibberish in the editor. But that's kinda LaTeX for ya. I'm not sure rendering it or giving other options for inserting would really help there, and it's mitigated some by the fact you need to know how to put it in in the first place, as opposed to having an "equations" button on the toolbar.

You _can_ break it by doing things like adding a link in the middle:

<img width="267" height="165" alt="Image" src="https://github.com/user-attachments/assets/d357fdb7-53ef-455c-8163-e48b77131818" />

but if the code is clean, the editor handles it well.

> Do we know if the ICON group is displaying that menu of options?

That's a good question that I'm not sure on. I did check, and was able to turn it off in configuration, so we could go either way here to be consistent.

**Comment by joewhitsitt** [comment]
>But that's kinda LaTeX for ya

Yeah, that is what I remember. I think it will be understood by those that need it.

**Comment by cory-skeers** [comment]
I mentioned in EoD, but want to note here:

In looking at the security implications here, MathJax seems the secure approach. It handles the rendering on client side, and we're not adding any LaTeX functionality in the system, so there's no direct access for some of LaTeX's invasive, direct commands to execute.

**Comment by cory-skeers** [comment]
Found one thing that doesn't always work:

Matrices use ampersands and double slashes. The ampersands get encoded, as expected for ckeditor, and that results in it being recognized by MathJax later in the filter order. The double slashes get reduced to a single, though, and so you lose the "next line" indicator.

<img width="452" height="130" alt="Image" src="https://github.com/user-attachments/assets/9dfcabaf-d5fc-417d-b3a8-8bf83e581f55" />

Should have been:

<img width="452" height="137" alt="Image" src="https://github.com/user-attachments/assets/96d33853-0829-4563-9b44-401fe72fabeb" />

Assuming the math-enabled text gets moved to a separate filter, there might be some filter adjustments that could be done to remedy this.

On the plus side, this is still the only instance I've come across where the encoding has negatively impacted it between editing and rendering. Who wants linear algebra, anyway?

**Comment by cory-skeers** [comment]
I think this was noted in EoD:

The MathType CKeditor plugin (https://ckeditor.com/mathtype/) requires a ckeditor premium upgrade or licensing through Wiris: https://www.wiris.com/en/mathtype/.

There is https://www.npmjs.com/package/ckeditor5-math as a nicely featured, open-source ckeditor plugin that adds a popup modal which allows rendering in the editor:

<img width="715" height="348" alt="Image" src="https://github.com/user-attachments/assets/69ec18c5-e247-47e3-b2e6-21432a0b38c4" />

Still looking some into this or if there are other alternatives, but so far:

Behind-the-scenes, we would likely configure it to use MathJax still instead of KaTeX, so the gain would be the better editor experience with no real difference on the end product, at the cost of extra custom infrastructure to load the plugin ourselves. (Maybe not. KaTeX doesn't support some LaTeX features that MathJax does, but it would take further investigating to see how much of a detriment that was).

**Comment by cory-skeers** [comment]
There's a POC with drupal/mathjax in https://github.com/uiowa/uiowa/pull/9802.

# Summary:

At the end of the day, KaTeX or non-MathJax implementations are probably a no-go. KaTeX has a client-side option, but some limitations on standard LaTeX options like `\label`s. I didn't really look into server-side implementations, as a) they're really involved to get set up and b) without careful setup, they introduce some pretty serious security concerns.

## MathJax
[drupal/mathjax ](https://www.drupal.org/project/mathjax) seems the way to go for the smallest lift on our end to get it up and running.

### Pros:
- Addition of one module
- "it just works" with CKEditor
- Server-side JS limits the security concerns

### Cons:
 - The editor is slightly clunky
 - Requires training and documentation
 - Isn't WYSIWYG
   - If it's used in layout builder, this is lessened, as once the block is saved, you can see the rendered output
 - Matrices can break unless the filter goes above the "fix faulty HTML" filter

## Direct CKEditor plugin:
After 4, CKEditor moved to a premium set for math which requires either their full premium editor or a license with Wiris
https://ckeditor.com/docs/ckeditor5/latest/features/math-equations.html

## [ckeditor-5-math](https://github.com/isaul32/ckeditor5-math#readme)
This is an open-source alternative for CKE5

### Pros:
 - Has a nice editor integration
 - Is more WYSIWYG with the above

### Cons:
 - We have to actually pull it in and include it
 - Requires extra infrastructure with dependencies that are not wrapped up in additional contrib modules

# Follow-up

There may be more cleanup and work to be done with the POC to get it ready for primetime, if that's the route we're going. There's also a need for documentation, either on its own, or possibly with the addition of help text which may be able to be included if its in its own "filtered html and also math" text format.

## Likely next steps

- Install drupal/mathjax (default? split?)
- Set up a new text format with mathjax (copy of filtered_html? more or less features? which roles have access?)
- Move the mathjax filter above the faulty html filter (or remove the faulty html filter for this format). Evaluate other filters.
- Evaluate and decide whether to turn on/off the math settings (see below)
- Documentation? Help text?
- Further testing

Math settings:
<img width="864" height="458" alt="Image" src="https://github.com/user-attachments/assets/e2b101df-3703-44b7-935a-1b92b8d762b1" />

**Comment by cory-skeers** [comment]
Closing the investigation as complete. Notes are in https://github.com/uiowa/uiowa/issues/9753#issuecomment-4392542886, and can follow-up with further planning as needed.
